crypto/cms/cms_sd.c

```diff
@@ -907,9 +913,16 @@ int CMS_SignerInfo_verify(CMS_SignerInfo *si)
         goto err;
     }
     mctx = si->mctx;
+    if (si->pctx != NULL) {
+        EVP_PKEY_CTX_free(si->pctx);
```

Comment:
Is this needed? Wont it just reuse it?

Reply:
Good question.
Yes it is needed, and no it wont reuse it.
The only way how `si->pctx` can be non-zero here,
is when `CMS_SignerInfo_verify is called twice.
But due to the `EVP_MD_CTX_reset(mctx);` at the end of this
function, both the flag `EVP_MD_CTX_FLAG_KEEP_PKEY_CTX`
and the reference inside the `mctx->pctx` is lost, and `si->pctx`
would create a memory leak, since the `EVP_DigestVerifyInit`
below will allocate a new pctx.  Unfortunately there are numerous
code paths that return the reference even in case of error.
I don't date to "fix" that, as it is only problematic together
with EVP_MD_CTX_FLAG_KEEP_PKEY_CTX.  Also setting that
flag before EVP_DigestVerifyInit is not a good idea, since
there are also code paths where the reference is not returned.